### ISSUE TITLE
AST diffing for first-time migrators

### ISSUE THREAD

**Issue by shaunlebron** [prettier/prettier#1552]
Are there any assurances that Prettier does not modify the AST?

Might be nice to check this for each test case, and additionally publish some API for assuring first-time migrators that their existing code bases won't change—this came up at Stripe.

I attempted to write such a script, but location meta and whitespace threw off the comparison:

```js
const fs = require('fs');
const prettier = require('prettier');
const stringify = require('json-stable-stringify');

const options = {};

const checkFile = (file) => {
  const text = fs.readFileSync(file, 'utf-8');
  const ast = prettier.__debug.parse(text, options);

  const prettierText = prettier.format(text, options);
  const prettierAst = prettier.__debug.parse(prettierText, options);

  // FIXME: this does not work due to location meta and whitespace inequalities
  const astJson = stringify(ast, {space: 2});
  const prettierAstJson = stringify(prettierAst, {space: 2});
  if (astJson != prettierAstJson) {
    throw new Error(`ast differs for ${file}`);
  }
};
```

We could use some prior work for selectively comparing ASTs by @roman01la here:
https://medium.com/@roman01la/compare-javascript-asts-ad8c94a2c701

**Comment by lydell** [comment]
Prettier does change the AST in one case:

```js
<p> This changes the AST.</p>
```
```js
<p>
  {" "}
  This changes the AST.
</p>;
```

So for an AST comparison tool to be truly effective, it would need to ignore this special case.

https://prettier.github.io/prettier/#%7B%22content%22%3A%22%3Cp%3E%20This%20changes%20the%20AST.%3C%2Fp%3E%22%2C%22options%22%3A%7B%22printWidth%22%3A20%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22none%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22semi%22%3Atrue%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%7D%7D

**Comment by bakkot** [comment]
I believe passing the undocumented `--debug-check` [flag](https://github.com/prettier/prettier/blob/44934da349e71db8a8a177c37bc368390286802b/bin/prettier.js#L141) will accomplish approximately that, including ignoring the above special case and a few others.

**Comment by shaunlebron** [comment]
Thanks!  `--debug-check` (#1337) works here for us.

I'll leave the issue open for a final question: should this option be renamed/documented for the use case of the first-time migrator?

**Comment by vjeux** [comment]
The goal of prettier is to be so good that you can trust it to do the right thing and it should never produce invalid code. We've been running extensive fuzz testing, ran it on a lot of giant codebases and treat any incorrect output as extremely high pri.

Now, this is indeed a scary proposition for people to run their codebase on prettier and we may want to expose it as a "peace of mind" rather than something we expect people to run and be afraid of the result.

**Comment by josephfrazier** [comment]
What if we add something to the end of the [CLI section of the README](https://github.com/prettier/prettier/tree/18fd0149859e98e051158db97dce321f8008c4a6#cli) along the lines of:

> If you're worried that prettier will change the correctness of your code, add `--debug-check` to the command. This will cause prettier to print an error message if it detects that code correctness might have changed.

Note that `prettier --write --debug-check` currently removes all content from the file if the `--debug-check` fails, so we'd probably want to have it do nothing instead, and also exit with a non-zero code to indicate failure.

**Comment by vjeux** [comment]
Sounds good, can you send a PR? :)

Can you also ensure that you cannot use both `--write` and `--debug-check` at the same time. `--debug-check` was meant to be used on its own, not with `--write`.

**Comment by bakkot** [comment]
I think this was closed by #1562, yes?

**Comment by josephfrazier** [comment]
Yeah, I probably should have closed this from a commit message there. At the time, I was thinking that we might want to change the wording of the message, but we can always re-open if that's an issue.
